### PR TITLE
Complete register gradient for compile time

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: paddle/framework/op_registry.cc [hunk 1]
```diff
@@ -52,9 +54,9 @@ std::unique_ptr<OperatorBase> OpRegistry::CreateOp(const OpDesc& op_desc) {
   return CreateOp(op_desc.type(), inputs, outputs, attrs);
 }
 
-std::unique_ptr<OperatorBase> OpRegistry::CreateGradOp(const OperatorBase& op) {
-  PADDLE_ENFORCE(!op.IsNetOp(), "Use framework::Backward to get backward ops");
-  return std::unique_ptr<OperatorBase>(BuildGradOp(&op));
+std::unique_ptr<OperatorBase> OpRegistry::CreateOp(OpDescBind* op_desc) {
```
Comment (Canpio) [inline paddle/framework/op_registry.cc]: `op_desc` should be a reference.

File: paddle/operators/mean_op.cc [hunk 2]
```diff
@@ -52,11 +52,27 @@ class MeanGradOp : public framework::OperatorWithKernel {
...
+ public:
+  using framework::SingleGradOpDescMaker::SingleGradOpDescMaker;
+
+ protected:
+  std::unique_ptr<framework::OpDescBind> Apply() const override {
+    auto* grad_op = new framework::OpDescBind();
+    grad_op->SetType("mean_grad");
+    grad_op->SetInput("X", Input("X"));
+    grad_op->SetInput(framework::GradVarName("Out"), OutputGrad("Out"));
+    grad_op->SetOutput(framework::GradVarName("X"), InputGrad("X"));
```
Comment (Canpio) [inline paddle/operators/mean_op.cc]: I'm adding a new constructor to `OpDescBind`, which initializes all its fields at one time.

```cpp
OpDescBind::OpDescBind(const std::string &type, const VariableNameMap &inputs,
                       const VariableNameMap &outputs,
                       const AttributeMap &attrs) {
  op_desc_.set_type(type);
  inputs_ = inputs;
  outputs_ = outputs;
  attrs_ = attrs;
}
```

File: paddle/operators/pad_op.cc [hunk 3]
```diff
@@ -111,11 +110,29 @@ class PadOpGrad : public framework::OperatorWithKernel {
   }
 };
 
+class PadOpGradMaker : public framework::SingleGradOpDescMaker {
+ public:
+  using framework::SingleGradOpDescMaker::SingleGradOpDescMaker;
+
+ protected:
+  std::unique_ptr<framework::OpDescBind> Apply() const override {
```
Comment (Canpio) [inline paddle/operators/pad_op.cc]: I still think `Apply()` is not necessary. We can let op developers override `operator()` directly by themselves. The definition of `operator()` has dictated that the return value should be a `vector`.

Comment (wangkuiyi) [reply paddle/operators/pad_op.cc]: We can discuss later and change if necessary.

File: paddle/framework/op_registry.h [hunk 4]
```diff
@@ -138,33 +137,41 @@ class OpKernelRegistrar : public Registrar {
                              __test_global_namespace_##uniq_name##__>::value, \
                 msg)
 
+#define REGISTER_OPERATOR(op_type, op_class, ...)                      \
+  STATIC_ASSERT_GLOBAL_NAMESPACE(                                      \
+      __reg_op__##op_type,                                             \
+      "REGISTER_OPERATOR must be called in global namespace");         \
+  class _OpClass_##op_type##_ : public op_class {                      \
+   public:                                                             \
+    DEFINE_OP_CLONE_METHOD(_OpClass_##op_type##_);                     \
```
[the remark this answers is not in the document]
Comment (wangkuiyi) [reply paddle/framework/op_registry.h]: Need discussion. Will change later if necessary.

[the remark this answers is not in the document]
Comment (reyoung) [reply paddle/framework/op_registry.h]: > Do we still need to make Op being able to copy themselves?

I think it is not needed.

**Pull request merged**